core/nodepool/config/config.go

```diff
@@ -0,0 +1,328 @@
+		if !s.HasIdentifier() {
+			stackOutputName := fmt.Sprintf(`{"Fn::ImportValue":{"Fn::Sub":"${ControlPlaneStackName}-%s"}}`, s.LogicalName())
+			az := s.AvailabilityZone
+			if s.Private {
+				c.Subnets[i] = model.NewPrivateSubnetFromFn(az, stackOutputName)
+			} else {
+				c.Subnets[i] = model.NewPublicSubnetFromFn(az, stackOutputName)
+			}
+		}
+	}
+
+	c.EtcdInstances = main.EtcdInstances
+
+	return nil
+}
+
+func ClusterFromBytesWithEncryptService(data []byte, main *cfg.Config, encryptService cfg.EncryptService) (*ProvidedConfig, error) {
+	cluster, err := ClusterFromBytes(data, main)
+	if err != nil {
+		return nil, err
+	}
+	cluster.providedEncryptService = encryptService
+	return cluster, nil
+}
+
+func (c ProvidedConfig) Config() (*ComputedConfig, error) {
+	config := ComputedConfig{ProvidedConfig: c}
+
+	if c.AmiId == "" {
+		var err error
+		if config.AMI, err = amiregistry.GetAMI(config.Region, config.ReleaseChannel); err != nil {
+			return nil, fmt.Errorf("failed getting AMI for config: %v", err)
+		}
+	} else {
+		config.AMI = c.AmiId
+	}
+
+	return &config, nil
+}
+
+func (c ProvidedConfig) WorkerDeploymentSettings() WorkerDeploymentSettings {
+	return WorkerDeploymentSettings{
+		WorkerNodePoolConfig: c.WorkerNodePoolConfig,
+		Experimental:         c.Experimental,
+		DeploymentSettings:   c.DeploymentSettings,
+	}
+}
+
+func (c ProvidedConfig) ValidateInputs() error {
+	if err := c.DeploymentSettings.ValidateInputs(); err != nil {
+		return err
+	}
+
+	if err := c.WorkerNodePoolConfig.ValidateInputs(); err != nil {
+		return err
+	}
+
+	if len(c.Subnets) > 1 && c.ClusterAutoscaler.Enabled() {
+		return errors.New("cluster-autoscaler support can't be enabled for a node pool with 2 or more subnets because allowing so" +
+			"results in unreliability while scaling nodes out. ")
+	}
+
+	return nil
+}
+
+func (c ProvidedConfig) valid() error {
+	if _, err := c.KubeClusterSettings.Valid(); err != nil {
+		return err
+	}
+
+	if err := c.WorkerNodePoolConfig.Validate(); err != nil {
+		return err
+	}
+
+	if err := c.DeploymentSettings.Valid(); err != nil {
+		return err
+	}
+
+	if err := c.WorkerDeploymentSettings().Valid(); err != nil {
+		return err
+	}
+
+	if err := c.Experimental.Valid(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// StackName returns the logical name of a CloudFormation stack resource in a root stack template
+// This is not needed to be unique in an AWS account because the actual name of a nested stack is generated randomly
+// by CloudFormation by including the logical name.
+// This is NOT intended to be used to reference stack name from cloud-config as the target of awscli or cfn-bootstrap-tools commands e.g. `cfn-init` and `cfn-signal`
+func (c ProvidedConfig) StackName() string {
+	return c.NodePoolName
+}
+
+func (c ProvidedConfig) StackNameEnvFileName() string {
+	return "/etc/environment"
+}
+
+func (c ProvidedConfig) StackNameEnvVarName() string {
+	return "KUBE_AWS_STACK_NAME"
```

Comment:
In other words, I thought it would be better than directly embedding the var name "KUBE_AWS_STACK_NAME" multiple places in cloud-config and stack templates ❤️